File: labs-opensource/M02-tokens-context-limits/solution/token_counting.py
"""
M02 Lab - Step 1: Token Counting with tiktoken — SOLUTION
==========================================================
Run: python token_counting.py
"""

import tiktoken

enc = tiktoken.get_encoding("cl100k_base")

# ── Part A: Basic token counting ──
samples = [
    "Hello, world!",
    "tokenization",
    "don't use pseudocode",
    "ChatGPT is one token",
    "from openai import OpenAI",
]

print("── Token breakdown ──")
for text in samples:
    ids = enc.encode(text)
    tokens = [enc.decode_single_token_bytes(t).decode("utf-8", errors="replace") for t in ids]
    print(f"{len(ids):3d} tokens | {tokens!r}")
    print(f"           text: {text!r}\n")


# ── Part B: Chat message overhead ──
def count_chat_tokens(messages: list[dict]) -> int:
    """Count tokens for a list of chat messages including overhead."""
    total = 3  # primer for assistant reply
    for msg in messages:
        total += 4  # role, content field markers, sep
        total += len(enc.encode(msg.get("role", "")))
        total += len(enc.encode(msg.get("content", "")))
    return total


messages = [
    {"role": "system", "content": "You are a helpful coding assistant."},
    {"role": "user", "content": "Explain what a token is in 2 sentences."},
]
print(f"Chat token estimate: {count_chat_tokens(messages)}")
# Output: Chat token estimate: ~31

File: labs-opensource/M02-tokens-context-limits/solution/token_counting.js
/**
 * M02 Lab - Step 1: Token Counting with js-tiktoken — SOLUTION
 * =============================================================
 * Run: node token_counting.js
 */

import { getEncoding } from "js-tiktoken";

const enc = getEncoding("cl100k_base");

// ── Part A: Basic token counting ──
const samples = [
  "Hello, world!",
  "tokenization",
  "don't use pseudocode",
  "ChatGPT is one token",
  "from openai import OpenAI",
];

console.log("── Token breakdown ──");
for (const text of samples) {
  const ids = enc.encode(text);
  console.log(`${String(ids.length).padStart(3)} tokens | "${text}"`);
}
console.log();

// ── Part B: Chat message overhead ──
function countChatTokens(messages) {
  let total = 3; // primer for assistant reply
  for (const msg of messages) {
    total += 4; // role + content markers + sep
    total += enc.encode(msg.role ?? "").length;
    total += enc.encode(msg.content ?? "").length;
  }
  return total;
}

const messages = [
  { role: "system", content: "You are a helpful coding assistant." },
  { role: "user", content: "Explain what a token is in 2 sentences." },
];
console.log(`Chat token estimate: ${countChatTokens(messages)}`);
// Output: Chat token estimate: ~31
